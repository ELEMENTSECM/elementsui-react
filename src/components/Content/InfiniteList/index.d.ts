import * as React from 'react';

export interface InfiniteListProps {
	/**
	 * DOM element's id attribute
	 */
	id?: string;
	/**
	 * Locale (en, nb, etc.)
	 */
	locale?: string;
	/**
	 * A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6].
	 */
	next?: (...args: any[]) => any;
	/**
	 * It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user
	 */
	hasMore?: boolean;
	/**
	 * The data items which you need to scroll.
	 */
	children?: React.ReactNode;
	/**
	 * This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false
	 */
	endMessage?: React.ReactNode;
	/**
	 * Optional, give only if you want to have a fixed height scrolling content
	 */
	height?: number;
	/**
	 * Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably.
	 */
	scrollableTarget?: React.ReactNode;
	/**
	 * Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more.
	 */
	hasChildren?: boolean;
	/**
	 * To enable Pull Down to Refresh feature
	 */
	pullDownToRefresh?: boolean;
	/**
	 * Any JSX that you want to show the user
	 */
	pullDownToRefreshContent?: React.ReactNode;
	/**
	 * Any JSX that you want to show the user
	 */
	releaseToRefreshContent?: React.ReactNode;
	/**
	 * Minimum distance the user needs to pull down to trigger the refresh
	 */
	pullDownToRefreshThreshold?: number;
	/**
	 * This function will be called, it should return the fresh data that you want to show the user
	 */
	refreshFunction?: (...args: any[]) => any;
	/**
	 * A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect.
	 */
	onScroll?: (...args: any[]) => any;
	/**
	 * Set the length of the data.This will unlock the subsequent calls to next.
	 */
	dataLength: number;
	/**
	 * User-defined styling
	 */
	styles?: (...args: any[]) => any;
}

declare const InfiniteList: React.SFC<InfiniteListProps>;

export default InfiniteList;
