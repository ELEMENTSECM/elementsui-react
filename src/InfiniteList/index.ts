export { default, InfiniteListProps } from "./InfiniteList";
export { default as ListItem, InfiniteListItemProps } from "./InfiniteListItem";
