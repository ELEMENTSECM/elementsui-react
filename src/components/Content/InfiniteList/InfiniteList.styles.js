export const styles = props => {
	const { className } = props;

	return {
		root: [
			'InfiniteList',
			{
				overflow: 'auto',
				WebkitOverflowScrolling: 'touch'
			},
			className
		],
		pulldown: [{ position: 'relative' }],
		pulldownHandle: [
			{
				position: 'absolute',
				left: 0,
				right: 0
			}
		],
		list: []
	};
};
