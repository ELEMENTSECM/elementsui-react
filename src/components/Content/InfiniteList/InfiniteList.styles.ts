import { css } from "react-emotion";

export const styles = {
	InfiniteList: css`
		overflow: auto;
		-webkit-overflow-scrolling: touch;
	`,
	pulldown: css`position: relative;`,
	pulldownHandle: css`
		position: absolute;
		,: 0;
		right: 0;
	`
};
