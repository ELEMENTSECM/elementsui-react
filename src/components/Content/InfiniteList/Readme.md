InfiniteList

```js
<InfiniteList locale="en" dataLength={20} hasMore={false}>
	{[ ...Array(20) ].map((_, i) => <div key={i}>Item {i}</div>)}
</InfiniteList>
```