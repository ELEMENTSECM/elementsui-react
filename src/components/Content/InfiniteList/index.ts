export { default, IInfiniteListProps } from "./InfiniteList";
