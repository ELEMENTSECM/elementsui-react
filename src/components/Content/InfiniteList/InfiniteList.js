import * as React from 'react';
import PropTypes from 'prop-types';
import { styles } from './InfiniteList.styles';
import { classNamesFunction, customizable, styled } from 'office-ui-fabric-react/lib/Utilities';
import Spinner from '../../Indicators/Spinner';
import { IntlProvider, FormattedMessage } from 'react-intl';
import nls from './InfiniteList.nls.json';

/** InfiniteList example */
class InfiniteListComponent extends React.Component {
	static getDerivedStateFromProps(props, state) {
		if (props.dataLength === state.dataLength) return null;

		return {
			dataLength: props.dataLength,
			showLoader: false,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};
	}

	constructor(props) {
		super();
		this.state = {
			dataLength: props.dataLength,
			showLoader: false,
			lastScrollTop: 0,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};

		this.startY = 0;
		this.currentY = 0;
		this.dragging = false;
		this.maxPullDownDistance = 0;

		this.throttledOnScrollListener = this.throttle(this.onScrollListener, 150);
		this.classNames = classNamesFunction()(styles, props);
	}

	componentDidMount() {
		this._scrollableNode = this.getScrollableTarget();
		this.el = this.props.height ? this._infScroll : this._scrollableNode || window;
		this.el.addEventListener('scroll', this.throttledOnScrollListener);

		if (
			typeof this.props.initialScrollY === 'number' &&
			this.el.scrollHeight > this.props.initialScrollY
		) {
			this.el.scrollTo(0, this.props.initialScrollY);
		}

		if (this.props.pullDownToRefresh) {
			this.el.addEventListener('touchstart', this.onStart);
			this.el.addEventListener('touchmove', this.onMove);
			this.el.addEventListener('touchend', this.onEnd);

			this.el.addEventListener('mousedown', this.onStart);
			this.el.addEventListener('mousemove', this.onMove);
			this.el.addEventListener('mouseup', this.onEnd);

			this.maxPullDownDistance = this._pullDown.firstChild.getBoundingClientRect().height;
			this.forceUpdate();

			if (typeof this.props.refreshFunction !== 'function') {
				throw new Error(
					`Mandatory prop "refreshFunction" missing.
					Pull Down To Refresh functionality will not work
					as expected.`
				);
			}
		}
	}

	componentWillUnmount() {
		this.el.removeEventListener('scroll', this.throttledOnScrollListener);

		if (this.props.pullDownToRefresh) {
			this.el.removeEventListener('touchstart', this.onStart);
			this.el.removeEventListener('touchmove', this.onMove);
			this.el.removeEventListener('touchend', this.onEnd);

			this.el.removeEventListener('mousedown', this.onStart);
			this.el.removeEventListener('mousemove', this.onMove);
			this.el.removeEventListener('mouseup', this.onEnd);
		}
	}

	getScrollableTarget = () => {
		if (this.props.scrollableTarget instanceof HTMLElement) return this.props.scrollableTarget;
		if (typeof this.props.scrollableTarget === 'string') {
			return document.getElementById(this.props.scrollableTarget);
		}
		if (this.props.scrollableTarget === null) {
			console.warn(`You are trying to pass scrollableTarget but it is null. This might
			happen because the element may not have been added to DOM yet.`);
		}
		return null;
	};

	onStart = evt => {
		if (this.state.lastScrollTop) return;

		this.dragging = true;
		this.startY = evt.pageY || evt.touches[0].pageY;
		this.currentY = this.startY;

		this._infScroll.style.willChange = 'transform';
		this._infScroll.style.transition = `transform 0.2s cubic-bezier(0,0,0.31,1)`;
	};

	onMove = evt => {
		if (!this.dragging) return;
		this.currentY = evt.pageY || evt.touches[0].pageY;

		if (this.currentY < this.startY) return;

		const threshold = this.props.pullDownToRefreshThreshold || 100;
		if (this.currentY - this.startY >= threshold) {
			this.setState(ps => ({
				...ps,
				pullToRefreshThresholdBreached: true
			}));
		}

		if (this.currentY - this.startY > this.maxPullDownDistance * 1.5) return;

		this._infScroll.style.overflow = 'visible';
		this._infScroll.style.transform = `translate3d(0px, ${this.currentY - this.startY}px, 0px)`;
	};

	onEnd = evt => {
		this.startY = 0;
		this.currentY = 0;

		this.dragging = false;

		if (this.state.pullToRefreshThresholdBreached) {
			this.props.refreshFunction && this.props.refreshFunction();
		}

		requestAnimationFrame(() => {
			if (this._infScroll) {
				this._infScroll.style.overflow = 'auto';
				this._infScroll.style.transform = 'none';
				this._infScroll.style.willChange = 'none';
			}
		});
	};

	isElementAtBottom(target, scrollThreshold = 0.8) {
		const clientHeight =
			target === document.body || target === document.documentElement
				? window.screen.availHeight
				: target.clientHeight;

		return target.scrollTop + clientHeight >= scrollThreshold * target.scrollHeight;
	}

	onScrollListener = event => {
		if (typeof this.props.onScroll === 'function') {
			setTimeout(() => this.props.onScroll(event), 0);
		}

		let target =
			this.props.height || this._scrollableNode
				? event.target
				: document.documentElement.scrollTop
					? document.documentElement
					: document.body;

		if (this.state.actionTriggered) return;

		let atBottom = this.isElementAtBottom(target, this.props.scrollThreshold);

		if (atBottom && this.props.hasMore) {
			this.setState(ps => ({ ...ps, actionTriggered: true, showLoader: true }));
			this.props.next();
		}
		this.setState(ps => ({ ...ps, lastScrollTop: target.scrollTop }));
	};

	throttle = (fn, threshhold, scope) => {
		threshhold || (threshhold = 250);
		let last, deferTimer;
		return function() {
			const context = scope || this;

			const now = +new Date(),
				args = arguments;
			if (last && now < last + threshhold) {
				clearTimeout(deferTimer);
				deferTimer = setTimeout(function() {
					last = now;
					fn.apply(context, args);
				}, threshhold);
			} else {
				last = now;
				fn.apply(context, args);
			}
		};
	};

	render() {
		const style = {
			height: this.props.height || 'auto'
		};
		const hasChildren = this.props.hasChildren || !!(this.props.children && this.props.children.length);

		const outerDivStyle = this.props.pullDownToRefresh && this.props.height ? { overflow: 'auto' } : {};
		const { root, pulldown, list, pulldownHandle } = this.classNames;
		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div style={outerDivStyle} className={root}>
					<div ref={infScroll => (this._infScroll = infScroll)} style={style}>
						{this.props.pullDownToRefresh && (
							<div className={pulldown} ref={pullDown => (this._pullDown = pullDown)}>
								<div
									className={pulldownHandle}
									style={{ top: -1 * this.maxPullDownDistance }}>
									{!this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="pullDownToRefresh" />
										</h3>
									)}
									{this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="releaseToRefresh" />
										</h3>
									)}
								</div>
							</div>
						)}
						<ul className={list}>{this.props.children}</ul>
						{!this.state.showLoader && !hasChildren && this.props.hasMore && <Spinner />}
						{this.state.showLoader && this.props.hasMore && <Spinner />}
						{!this.props.hasMore && this.props.endMessage}
					</div>
				</div>
			</IntlProvider>
		);
	}
}

export const InfiniteList = InfiniteListComponent;

InfiniteList.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

InfiniteListComponent.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string.isRequired,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

export default styled(customizable('InfiniteList', ['theme'])(InfiniteList), styles);
