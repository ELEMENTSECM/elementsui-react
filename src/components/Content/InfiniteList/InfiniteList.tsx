import * as React from "react";
import { styles } from "./InfiniteList.styles";
import Spinner from "../../Indicators/Spinner";
import { IntlProvider, FormattedMessage } from "react-intl";
import { nls } from "./InfiniteList.nls";

export type InfiniteListHeight = number | string;

export interface IInfiniteListProps {
	/**
	 * DOM element's id attribute
	 */
	id?: string;
	/**
	 * Locale (en, nb, etc.)
	 */
	locale?: string;
	/**
	 * A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6].
	 */
	next?: (...args: any[]) => any;
	/**
	 * It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user
	 */
	hasMore?: boolean;
	/**
	 * The data items which you need to scroll.
	 */
	children?: React.ReactNode;
	/**
	 * This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false
	 */
	endMessage?: React.ReactNode;
	/**
	 * Optional, give only if you want to have a fixed height scrolling content
	 */
	height?: InfiniteListHeight;
	/**
	 * Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably.
	 */
	scrollableTarget?: React.ReactNode;
	/**
	 * Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more.
	 */
	hasChildren?: boolean;
	/**
	 * To enable Pull Down to Refresh feature
	 */
	pullDownToRefresh?: boolean;
	/**
	 * Any JSX that you want to show the user
	 */
	pullDownToRefreshContent?: React.ReactNode;
	/**
	 * Any JSX that you want to show the user
	 */
	releaseToRefreshContent?: React.ReactNode;
	/**
	 * Minimum distance the user needs to pull down to trigger the refresh
	 */
	pullDownToRefreshThreshold?: number;
	/**
	 * This function will be called, it should return the fresh data that you want to show the user
	 */
	refreshFunction?: (...args: any[]) => any;
	/**
	 * A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect.
	 */
	onScroll?: (...args: any[]) => any;
	/**
	 * Set the length of the data.This will unlock the subsequent calls to next.
	 */
	dataLength: number;
	/**
	 * Root div class name
	 */
	className?: string;
	/**
	 * List class name
	 */
	listClassName?: string;
}

interface IState {
	dataLength: number;
	showLoader: boolean;
	lastScrollTop: number;
	actionTriggered: boolean;
	pullToRefreshThresholdBreached: boolean;
}

/** InfiniteList example */
export default class InfiniteList extends React.Component<IInfiniteListProps, IState> {
	static getDerivedStateFromProps(props: IInfiniteListProps, state: IState) {
		if (props.dataLength === state.dataLength) {
			return null;
		}

		return {
			dataLength: props.dataLength,
			showLoader: false,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};
	}

	private startY: number;
	private currentY: number;
	private dragging: boolean;
	private maxPullDownDistance: number;
	private throttledOnScrollListener: () => any;
	private infScroll: HTMLDivElement;
	private el: HTMLElement | Window;
	private scrollableNode: HTMLElement;
	private pullDown: HTMLDivElement;

	constructor(props: IInfiniteListProps) {
		super(props);
		this.state = {
			dataLength: props.dataLength,
			showLoader: false,
			lastScrollTop: 0,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};

		this.startY = 0;
		this.currentY = 0;
		this.dragging = false;
		this.maxPullDownDistance = 0;

		this.throttledOnScrollListener = this.throttle(this.onScrollListener, 150);
	}

	componentDidMount() {
		this.scrollableNode = this.getScrollableTarget();
		this.el = this.props.height ? this.infScroll : this.scrollableNode || window;
		this.el.addEventListener("scroll", this.throttledOnScrollListener);

		if (this.props.pullDownToRefresh) {
			this.el.addEventListener("touchstart", this.onStart);
			this.el.addEventListener("touchmove", this.onMove);
			this.el.addEventListener("touchend", this.onEnd);

			this.el.addEventListener("mousedown", this.onStart);
			this.el.addEventListener("mousemove", this.onMove);
			this.el.addEventListener("mouseup", this.onEnd);

			this.maxPullDownDistance = this.pullDown.firstChild["getBoundingClientRect"]().height;
			this.forceUpdate();

			if (typeof this.props.refreshFunction !== "function") {
				throw new Error(
					`Mandatory prop "refreshFunction" missing.
					Pull Down To Refresh functionality will not work
					as expected.`
				);
			}
		}
	}

	componentWillUnmount() {
		this.el.removeEventListener("scroll", this.throttledOnScrollListener);

		if (this.props.pullDownToRefresh) {
			this.el.removeEventListener("touchstart", this.onStart);
			this.el.removeEventListener("touchmove", this.onMove);
			this.el.removeEventListener("touchend", this.onEnd);

			this.el.removeEventListener("mousedown", this.onStart);
			this.el.removeEventListener("mousemove", this.onMove);
			this.el.removeEventListener("mouseup", this.onEnd);
		}
	}

	getScrollableTarget = () => {
		if (this.props.scrollableTarget instanceof HTMLElement) {
			return this.props.scrollableTarget;
		}
		if (typeof this.props.scrollableTarget === "string") {
			return document.getElementById(this.props.scrollableTarget);
		}
		if (this.props.scrollableTarget === null) {
			console.warn(`You are trying to pass scrollableTarget but it is null. This might
			happen because the element may not have been added to DOM yet.`);
		}
		return null;
	};

	onStart = (evt) => {
		if (this.state.lastScrollTop) {
			return;
		}

		this.dragging = true;
		this.startY = evt.pageY || evt.touches[0].pageY;
		this.currentY = this.startY;

		this.infScroll.style["willChange"] = "transform";
		this.infScroll.style.transition = `transform 0.2s cubic-bezier(0,0,0.31,1)`;
	};

	onMove = (evt) => {
		if (!this.dragging) {
			return;
		}
		this.currentY = evt.pageY || evt.touches[0].pageY;

		if (this.currentY < this.startY) {
			return;
		}

		const threshold = this.props.pullDownToRefreshThreshold || 100;
		if (this.currentY - this.startY >= threshold) {
			this.setState((ps) => ({
				...ps,
				pullToRefreshThresholdBreached: true
			}));
		}

		if (this.currentY - this.startY > this.maxPullDownDistance * 1.5) {
			return;
		}

		this.infScroll.style.overflow = "visible";
		this.infScroll.style.transform = `translate3d(0px, ${this.currentY - this.startY}px, 0px)`;
	};

	onEnd = () => {
		this.startY = 0;
		this.currentY = 0;

		this.dragging = false;

		if (this.state.pullToRefreshThresholdBreached) {
			this.props.refreshFunction && this.props.refreshFunction();
		}

		requestAnimationFrame(() => {
			if (this.infScroll) {
				this.infScroll.style.overflow = "auto";
				this.infScroll.style.transform = "none";
				this.infScroll.style["willChange"] = "none";
			}
		});
	};

	isElementAtBottom(target: any) {
		const clientHeight =
			target === document.body || target === document.documentElement
				? window.screen.availHeight
				: target.clientHeight;

		return target.scrollHeight - target.scrollTop === clientHeight;
	}

	onScrollListener = (event) => {
		if (typeof this.props.onScroll === "function") {
			setTimeout(() => this.props.onScroll(event), 0);
		}

		const target =
			this.props.height || this.scrollableNode
				? event.target
				: document.documentElement.scrollTop ? document.documentElement : document.body;

		if (this.state.actionTriggered) {
			return;
		}

		const atBottom = this.isElementAtBottom(target);

		if (atBottom && this.props.hasMore) {
			this.setState((ps) => ({ ...ps, actionTriggered: true, showLoader: true }));
			this.props.next();
		}
		this.setState((ps) => ({ ...ps, lastScrollTop: target.scrollTop }));
	};

	throttle = (fn, threshhold, scope?) => {
		threshhold || (threshhold = 250);
		let last, deferTimer;
		return function() {
			const context = scope || this;

			const now = +new Date(),
				args = arguments;
			if (last && now < last + threshhold) {
				clearTimeout(deferTimer);
				deferTimer = setTimeout(() => {
					last = now;
					fn.apply(context, args);
				}, threshhold);
			} else {
				last = now;
				fn.apply(context, args);
			}
		};
	};

	render() {
		const style = {
			height: this.props.height || "auto"
		};
		const hasChildren = this.props.hasChildren || !!this.props.children;
		const outerDivStyle = this.props.pullDownToRefresh && this.props.height ? { overflow: "auto" } : {};
		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div
					style={Object.assign(style, outerDivStyle, styles.InfiniteList)}
					className={`${styles.InfiniteList} ${this.props.className}`}
					ref={(infScroll) => (this.infScroll = infScroll)}
				>
					{this.props.pullDownToRefresh && (
						<div className={styles.pulldown} ref={(pullDown) => (this.pullDown = pullDown)}>
							<div className={styles.pulldownHandle} style={{ top: -1 * this.maxPullDownDistance }}>
								{!this.state.pullToRefreshThresholdBreached && (
									<h3>
										<FormattedMessage id="pullDownToRefresh" />
									</h3>
								)}
								{this.state.pullToRefreshThresholdBreached && (
									<h3>
										<FormattedMessage id="releaseToRefresh" />
									</h3>
								)}
							</div>
						</div>
					)}
					<ul className={this.props.listClassName}>{this.props.children}</ul>
					{!this.state.showLoader && !hasChildren && this.props.hasMore && <Spinner size={50} />}
					{this.state.showLoader && this.props.hasMore && <Spinner size={50} />}
					{!this.props.hasMore && this.props.endMessage}
				</div>
			</IntlProvider>
		);
	}
}
