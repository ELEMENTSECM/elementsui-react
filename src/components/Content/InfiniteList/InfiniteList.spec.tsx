import * as React from "react";
import { shallow } from "enzyme";
import InfiniteList from "./index";

describe("InfiniteList", () => {
	test("displays correctly", () => {
		const wrapper = shallow(
			<InfiniteList locale="en" dataLength={20} hasMore={false}>
				{[ ...Array(20) ].map((_, i) => {
					return <div key={i}>Item {i}</div>;
				})}
			</InfiniteList>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
