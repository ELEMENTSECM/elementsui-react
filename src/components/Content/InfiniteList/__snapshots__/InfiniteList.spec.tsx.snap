// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteList displays correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InfiniteList
    dataLength={20}
    hasMore={false}
    locale="en"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div
        className="css-apz85e undefined"
        style={
          Object {
            "0": "c",
            "1": "s",
            "2": "s",
            "3": "-",
            "4": "a",
            "5": "p",
            "6": "z",
            "7": "8",
            "8": "5",
            "9": "e",
            "height": "auto",
          }
        }
      >
        <ul
          className={undefined}
        />
      </div>,
      "locale": "en",
      "messages": Object {
        "pullDownToRefresh": "Pull down to refresh",
        "releaseToRefresh": "Release to refresh",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <ul
            className={undefined}
          />,
          false,
          false,
          undefined,
        ],
        "className": "css-apz85e undefined",
        "style": Object {
          "0": "c",
          "1": "s",
          "2": "s",
          "3": "-",
          "4": "a",
          "5": "p",
          "6": "z",
          "7": "8",
          "8": "5",
          "9": "e",
          "height": "auto",
        },
      },
      "ref": [Function],
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
        false,
        false,
        undefined,
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          className="css-apz85e undefined"
          style={
            Object {
              "0": "c",
              "1": "s",
              "2": "s",
              "3": "-",
              "4": "a",
              "5": "p",
              "6": "z",
              "7": "8",
              "8": "5",
              "9": "e",
              "height": "auto",
            }
          }
        >
          <ul
            className={undefined}
          />
        </div>,
        "locale": "en",
        "messages": Object {
          "pullDownToRefresh": "Pull down to refresh",
          "releaseToRefresh": "Release to refresh",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <ul
              className={undefined}
            />,
            false,
            false,
            undefined,
          ],
          "className": "css-apz85e undefined",
          "style": Object {
            "0": "c",
            "1": "s",
            "2": "s",
            "3": "-",
            "4": "a",
            "5": "p",
            "6": "z",
            "7": "8",
            "8": "5",
            "9": "e",
            "height": "auto",
          },
        },
        "ref": [Function],
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ],
              "className": undefined,
            },
            "ref": null,
            "rendered": Array [],
            "type": "ul",
          },
          false,
          false,
          undefined,
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
